server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -445,7 +445,13 @@ private static void parseObject(final DocumentParserContext context, ObjectMappe
         Mapper objectMapper = getMapper(context, mapper, currentFieldName);
         if (objectMapper != null) {
             context.path().add(currentFieldName);
+            if (objectMapper instanceof ObjectMapper objMapper) {
+                if (objMapper.isCollapsed()) {
+                    context.path().setWithinCollapsedPath(true);
+                }
+            }
             parseObjectOrField(context, objectMapper);
+            context.path().setWithinCollapsedPath(false);
             context.path().remove();
```

Comment:
I think we can always set within collapsed path to false on `remove()` can't we, because by definition if we're moving back up the object hierarchy then the object we're moving to can't be collapsed.  So maybe we can have just one method on ContentPath, `setCollapsed()`?